lightning/api_grpc.go

```diff
@@ -790,7 +783,256 @@ func (l *LndGrpcLightningAPI) SubscribeFailedForwards(ctx context.Context, outch
 	return nil
 }
 
-// GetAPIType API
+// GetAPIType API.
 func (l *LndGrpcLightningAPI) GetAPIType() APIType {
 	return LndGrpc
 }
+
+// ConnectPeer API.
+func (l *LndGrpcLightningAPI) ConnectPeer(ctx context.Context, id string) error {
+	split := strings.Split(id, "@")
+	if len(split) != 2 {
+		return fmt.Errorf("invalid id")
+	}
+
+	_, err := l.Client.ConnectPeer(ctx, &lnrpc.ConnectPeerRequest{
+		Addr:    &lnrpc.LightningAddress{Host: split[1], Pubkey: split[0]},
+		Perm:    false,
+		Timeout: 10,
+	})
+	if err != nil {
+		if strings.Contains(err.Error(), "already connected to peer") {
+			err = nil
+			return err
+		}
+		return err
+	}
+
+	return nil
+}
+
+// GetOnChainAddress API.
+func (l *LndGrpcLightningAPI) GetOnChainAddress(ctx context.Context) (string, error) {
+	resp, err := l.Client.NewAddress(ctx, &lnrpc.NewAddressRequest{Type: lnrpc.AddressType_WITNESS_PUBKEY_HASH})
+
+	if err != nil {
+		return "", err
+	}
+
+	return resp.Address, nil
+}
+
+// GetOnChainFunds API.
+func (l *LndGrpcLightningAPI) GetOnChainFunds(ctx context.Context) (*Funds, error) {
+	resp, err := l.Client.WalletBalance(ctx, &lnrpc.WalletBalanceRequest{})
+	if err != nil {
+		return nil, err
+	}
+
+	f := &Funds{}
+	f.ConfirmedBalance = resp.ConfirmedBalance
+	f.TotalBalance = resp.TotalBalance
+	f.LockedBalance = resp.ReservedBalanceAnchorChan + resp.LockedBalance
+
+	return f, nil
+}
+
+// SendToOnChainAddress API.
+func (l *LndGrpcLightningAPI) SendToOnChainAddress(ctx context.Context, address string, sats int64, useUnconfirmed bool, urgency Urgency) (string, error) {
+	target := 1
+	switch urgency {
+	case Urgent:
+		target = 1
+	case Normal:
+		target = 4
+	case Low:
+		target = 100
+	}
+
+	resp, err := l.Client.SendCoins(ctx, &lnrpc.SendCoinsRequest{
+		Addr:             address,
+		Amount:           sats,
+		TargetConf:       int32(target),
+		SpendUnconfirmed: useUnconfirmed,
+	})
+
+	if err != nil {
+		return "", err
+	}
+
+	if len(resp.Txid) == 0 {
+		return "", fmt.Errorf("invalid transaction id")
+	}
+
+	return resp.Txid, nil
+}
+
+// PayInvoice API.
+func (l *LndGrpcLightningAPI) PayInvoice(ctx context.Context, paymentRequest string, sats int64, outgoingChanIds []uint64) (*PaymentResp, error) {
+	req := &routerrpc.SendPaymentRequest{}
+	req.PaymentRequest = paymentRequest
+	if sats > 0 {
+		req.Amt = sats
+	}
+
+	if outgoingChanIds != nil {
+		req.OutgoingChanIds = make([]uint64, 0)
+		req.OutgoingChanIds = append(req.OutgoingChanIds, outgoingChanIds...)
+	}
+
+	// TODO: this is mandatory field but timeout could be configurable
+	req.TimeoutSeconds = 10
+
+	resp, err := l.RouterClient.SendPaymentV2(ctx, req)
+
+	if err != nil {
+		if strings.Contains(err.Error(), "invoice is already paid") {
+			// TODO: we don't know the hash here, else we could return l.GetPaymentStatus()
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("timeout")
+		default:
```

Comment:
I really dont understand this pattern

looks like you dont need select

```
 if ctx.Err() != nil {
    return nil, fmt.Errorf("timeout")
 }
 ```

Reply:
```
// If Done is not yet closed, Err returns nil.
// If Done is closed, Err returns a non-nil error explaining why:
// Canceled if the context was canceled
// or DeadlineExceeded if the context's deadline passed.
// After Err returns a non-nil error, successive calls to Err return the same error.
```

Looks like you're right 😊  I just invented an ugly antipattern, let's fix this.